### PR TITLE
Hosted Article - 'more from' carousel

### PR DESCRIPTION
**Pull request by lps88**
## What does this change?

Add new functionality to the 'More from' component, so the user can scroll through more of the campaign's other pages. Currently switched off until both new variants are finished.
## What is the value of this and can you measure success?

There will be an AB test to measure whether this works better (i.e. gets more people to click through) than the current design.
## Screenshots

image
## Request for comment

@guardian/labs-beta

### CODE REVIEW

File: commercial/conf/routes [hunk 1]
```diff
@@ -49,7 +49,7 @@ GET         /commercial/paid                                              commer
 # Hosted content
 GET         /commercial/advertiser-content/:campaignName/:pageName          commercial.controllers.HostedContentController.renderLegacyHostedPage(campaignName, pageName)
 GET         /advertiser-content/:campaignName/:pageName                     commercial.controllers.HostedContentController.renderHostedPage(campaignName, pageName)
-GET         /advertiser-content/:campaignName/:pageName/:cType/onward.json  commercial.controllers.HostedContentController.renderOnwardComponent(campaignName, pageName, cType)
+GET         /advertiser-content/:campaignName/:pageName/:cType/onward.json  commercial.controllers.HostedContentController.renderOnwardComponent(campaignName, pageName, cType, items: Option[Int])
```
Comment (kelvin-chappell) [inline commercial/conf/routes]: Could get rid of new `items` param.

File: static/src/javascripts/projects/common/modules/experiments/tests/hosted-more-from.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        this.showForSensitive = true;
+        this.audience = 0;
+        this.audienceOffset = 0;
+        this.successMeasure = 'People will click on links from either the first or second variant';
+        this.audienceCriteria = '';
+        this.dataLinkNames = '';
+        this.idealOutcome = 'People will click on links from either the first or second variant';
+
+        this.canRun = function () {
+            return /*config.isHosted && */config.page.contentType === 'Gallery';
```
Comment (kelvin-chappell) [inline static/src/javascripts/projects/common/modules/experiments/tests/hosted-more-from.js]: Can this commented out condition be removed?

File: static/src/javascripts/projects/commercial/modules/hosted/onward.js [hunk 3]
```diff
@@ -13,16 +14,19 @@ define([
         var placeholders = document.querySelectorAll('.js-onward-placeholder');
 
         if (placeholders.length) {
+            var contentType = config.page.contentType.toLowerCase();
+            var query = contentType === 'article' ? '?items=12' : '';
```
Comment (kelvin-chappell) [inline static/src/javascripts/projects/commercial/modules/hosted/onward.js]: This becomes redundant if number of items is set in controller.

File: commercial/app/controllers/HostedContentController.scala [hunk 4]
```diff
@@ -84,13 +84,13 @@ class HostedContentController(contentApiClient: ContentApiClient)
           val itemId = s"advertiser-content/$campaignName/$pageName"
           contentType match {
             case "video" =>
-              val trails = HostedTrails.fromContent(itemId, trailCount = 1, results)
+              val trails = HostedTrails.fromContent(itemId, trailCount = items.getOrElse(1), results)
               Cached(cacheDuration)(JsonComponent(hostedVideoOnward(trails.headOption)))
             case "article" =>
-              val trails = HostedTrails.fromContent(itemId, trailCount = 2, results)
+              val trails = HostedTrails.fromContent(itemId, trailCount = items.getOrElse(2), results)
```
Comment (kelvin-chappell) [inline commercial/app/controllers/HostedContentController.scala]: Could change this to `trailCount = <large number>` and then get rid of new `items` param.

File: static/src/javascripts/projects/common/modules/experiments/tests/hosted-more-from.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+                }
+            },
+            {
+                id: 'variant2',
+                test: function () {
+                }
+            }
+        ];
+    };
+});
```
Comment (kelvin-chappell) [inline static/src/javascripts/projects/common/modules/experiments/tests/hosted-more-from.js]: This file should probably be put in a separate PR.

File: static/src/javascripts/projects/common/modules/experiments/ab.js [hunk 6]
```diff
@@ -25,14 +26,16 @@ define([
...
     var TESTS = [
         new DiscussionPromoteBottomBanner(),
         new AdBlockingResponse(),
         new WeekendReadingEmail(),
-        new WeekendReadingPromo()
+        new WeekendReadingPromo(),
+        new HostedMoreFrom()
     ].concat(MembershipEngagementBannerTests);
 
     var participationsKey = 'gu.ab.participations';
```
Comment (kelvin-chappell) [inline static/src/javascripts/projects/common/modules/experiments/ab.js]: This file should probably be in a separate PR.

File: common/app/conf/switches/ABTestSwitches.scala [hunk 7]
```diff
@@ -66,4 +66,13 @@ trait ABTestSwitches {
...
 
+  Switch(
+    ABTests,
+    "ab-hosted-more-from",
+    "Show more onward journey links in the hosted article pages",
+    owners = Seq(Owner.withGithub("Calanthe")),
+    safeState = Off,
+    sellByDate = new LocalDate(2016, 10, 27),
+    exposeClientSide = true
+  )
```
Comment (kelvin-chappell) [inline common/app/conf/switches/ABTestSwitches.scala]: This switch should probably be in a separate PR.

File: commercial/app/controllers/HostedContentController.scala (hunk 4 above)
Comment (kelvin-chappell) [inline commercial/app/controllers/HostedContentController.scala]: And then building hosted trails becomes a lot simpler so could be done in a different def, without specifying a trail count at all.  But that could be done in a separate PR.

**Review by kelvin-chappell — CHANGES_REQUESTED** [review]
Looks good but some of the code is redundant or not obviously related.

File: dev-build/conf/routes [hunk 8]
```diff
@@ -307,7 +307,7 @@ GET            /commercial/paid.json
 GET            /commercial/paid                                                                                                  commercial.controllers.PaidContentCardController.cardHtml
 GET            /commercial/multi.json                                                                                            commercial.controllers.Multi.renderMulti
 GET            /commercial/api/multi.json                                                                                        commercial.controllers.Multi.getMulti
-GET            /$path<commercial-containers>                                                                                     controllers.FaciaController.renderFront(path)
+GET            /$path<commercial-containers>                                                                                                controllers.FaciaController.renderFront(path)
```
Comment (kelvin-chappell) [inline dev-build/conf/routes]: Looks the same to me.

**Review by kelvin-chappell — APPROVED** [review]

File: dev-build/conf/routes (hunk 8 above)
Comment (lps88) [inline dev-build/conf/routes]: I think I accidentally committed some whitespace changes...!

**Pull request merged**